### PR TITLE
Implement Entry backend

### LINKED ISSUE #7: Entry concept backend
Remaining:
- [x] Date parsing & validation (basic ver. using `Date` casting implemented)
- [x] Type in set validation? (frontend dropdown + middleware validation; not in data model)
- [x] Relationship with Trend/Scale (??) --> add Condition & Scale to Entry data model?
- [x] Get rid of procedure in types since there's no syncing with other concepts
- [x] Allow setting specific time to entry? is that necessary complexity tho... --> revisit in final

### PR DESCRIPTION
**Pull request by wli-linda**
Closes #7

### CODE REVIEW

File: server/entry/collection.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  static async findOne(entryId: Types.ObjectId | string): Promise<HydratedDocument<Entry>> {
+    return EntryModel.findOne({_id: entryId}).populate(['owner']);
+  }
+
+  static async findAllByOwnerId(ownerId: Types.ObjectId | string): Promise<HydratedDocument<Entry>[]> {
+    return EntryModel.find({owner: ownerId}).sort({date: -1}).populate(['owner']);
+  }
+
+  static async updateOne(entryId: Types.ObjectId | string, entryDetails: {type?: string; detail?: string; notes?: string, date?: Date}): Promise<HydratedDocument<Entry>> {
```
Comment (wli-linda) [inline server/entry/collection.ts]: Should the type of `date` be `Date`, assuming that it's converted into `Date` in `router.ts`?

File: server/entry/collection.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    const entry = await EntryModel.findOne({_id: entryId});
+    if (entryDetails.type) {
+      entry.type = entryDetails.type;
+    }
+
+    if (entryDetails.detail) {
+      entry.detail = entryDetails.detail;
+    }
```
Comment (wli-linda) [inline server/entry/collection.ts:38]: Should there be some sync between updating `type` and `detail`, s.t. updating `type` must also require updating `detail`? Or is that a final problem not a beta one?

File: server/entry/middleware.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+const isValidEntryDetail = async (req: Request, res: Response, next: NextFunction) => {
+    const {type, detail} = req.body as {type: string, detail: string};
+    if (!type.trim || !detail.trim) {
+        res.status(400).json({
+            error: 'Entry type/detail must be at least one character long.'
+        });
+        return;
+    }
+
+    // TODO: check if type is in set of specified types
```
Comment (wli-linda) [inline server/entry/middleware.ts]: This is me not knowing where to provide the set of types in the backend

File: server/entry/middleware.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    next();
+}
+
+const isValidEntryDate = async (req: Request, res: Response, next: NextFunction) => {
+    // TODO
```
Comment (wli-linda) [inline server/entry/middleware.ts]: Should the date in the http request be a string, which is cast to `Date` here? Not familiar with UNIX timestamp manipulation in TS

File: server/entry/middleware.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            error: `Entry with entry ID ${entryId} does not exist.`
+        });
+        return;
+    }
+
+    next();
+}
+
+const isValidEntryDetail = async (req: Request, res: Response, next: NextFunction) => {
+    const {type, detail} = req.body as {type: string, detail: string};
```
Comment (wli-linda) [inline server/entry/middleware.ts:20]: Not sure if syntactically correct

File: server/entry/middleware.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    next();
+}
+
+const isValidEntryModifier = async (req: Request, res: Response, next: NextFunction) => {
+    const entry = await EntryCollection.findOne(req.params.entryId);
+    const accountId = entry.owner._id;
+    if (req.session.accountId !== accountId.toString()) {
+      res.status(403).json({
+        error: 'Cannot modify other users\' entries.'
```
Comment (wli-linda) [inline server/entry/middleware.ts]: *accounts\'

File: server/entry/model.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  _id: Types.ObjectId;
+  owner: U;
+  type: T;
+  detail: D;
+  notes: string;
+  date: Date;
+};
+
+export type Entry = ParameterizedEntry<Types.ObjectId, string, string>;
+export type PopulatedEntry = ParameterizedEntry<Account, string, string>;
```
Comment (wli-linda) [inline server/entry/model.ts]: Would it be easier if I just created some sort of enum type `Type`? Also, map each `Type` to `Medication`, `Contact` etc.?

File: server/entry/router.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+ *
+ * */
+router.patch(
+  '/:entryId',
+  [
+    accountValidator.isAccountLoggedIn,
+    entryValidator.isValidEntry,
+    entryValidator.isValidEntryModifier,
+    entryValidator.isValidEntryDetail,
+    entryValidator.isValidEntryDate,
```
Comment (wli-linda) [inline server/entry/router.ts]: This might be problematic, since in updating they're optional. Perhaps add check for `!== undefined` in middleware?

**Comment by wli-linda** [comment]
Also, I got a bit lost at the distinction between Entry & Trend concepts. I believe each Entry should be associated with a Condition and scale (1-5), independent of the Trend concept?

File: server/entry/model.ts (hunk 7 above)
Comment (wli-linda) [reply server/entry/model.ts]: Revisit in final

File: server/entry/router.ts (hunk 8 above)
Comment (wli-linda) [reply server/entry/router.ts]: Hmm, or actually if we assume default values are provided in the frontend when editing (i.e. the current values pre-edit), then this check should be fine?

**Review by oliviagiandrea — DISMISSED** [review]
everything looks good to me - great job with the middleware!

File: server/entry/router.ts (hunk 8 above)
Comment (oliviagiandrea) [reply server/entry/router.ts]: i think assuming default values come through the frontend is good - the way it is now seems to work well

**Review by oliviagiandrea — DISMISSED** [review]
fixed date validation with regex :3

**Review by oliviagiandrea — APPROVED** [review]
merge looks good

**Pull request merged**